pytensor/xtensor/type.py

```diff
@@ -0,0 +1,250 @@
+
+        return res
+
+    def __hash__(self):
+        return hash(super().__hash__(), self.shape, self.dims)
+
+    def __repr__(self):
+        # TODO: Add `?` for unknown shapes like `TensorType` does
+        return f"XTensorType({self.dtype}, {self.dims}, {self.shape})"
+
+    def __eq__(self, other):
+        res = super().__eq__(other)
+
+        if isinstance(res, bool):
+            return res and self.dims == other.dims and self.shape == other.shape
+
+        return res
+
+    def is_super(self, otype):
+        # TODO: Implement this
+        return True
+
+        if not super().is_super(otype):
+            return False
+
+        if self.dims == otype.dims:
+            return True
+
+        return False
+
+
+def xtensor(
+    name: str | None = None,
+    *,
+    dims: Sequence[str],
+    shape: Sequence[int | None] | None = None,
+    dtype: str | np.dtype = "floatX",
+):
+    return XTensorType(dtype, dims=dims, shape=shape, name=name)()
+
+
+# class _x_tensor_py_operators
+
+
+class XTensorVariable(Variable):
+    pass
+
+    # def __str__(self):
+    #     return f"{self.__class__.__name__}{{{self.format},{self.dtype}}}"
+
+    # def __repr__(self):
+    #     return str(self)
+
+
+class XTensorConstantSignature(tuple):
+    def __eq__(self, other):
+        if type(self) is not type(other):
+            return False
+
+        (t0, d0), (t1, d1) = self, other
+        if t0 != t1 or d0.shape != d1.shape:
+            return False
+
+        return True
+
+    def __ne__(self, other):
+        return not self == other
+
+    def __hash__(self):
+        (a, b) = self
+        return hash(type(self)) ^ hash(a) ^ hash(type(b))
+
+    def pytensor_hash(self):
+        t, d = self
+        return "".join([hash_from_ndarray(d)] + [hash_from_code(dim) for dim in t.dims])
+
+
+_XTensorTypeType = TypeVar("_XTensorTypeType", bound=XTensorType)
+
+
+class XTensorConstant(XTensorVariable, Constant[_XTensorTypeType]):
+    def __init__(self, type: _XTensorTypeType, data, name=None):
+        # TODO: Add checks that type and data are compatible
+        Constant.__init__(self, type, data, name)
+
+    def signature(self):
+        assert self.data is not None
+        return XTensorConstantSignature((self.type, self.data))
+
+
+XTensorType.variable_type = XTensorVariable
+XTensorType.constant_type = XTensorConstant
+
+
+def xtensor_constant(x, name=None):
+    if not isinstance(x, xr.DataArray):
+        raise TypeError("xtensor.constant must be called on a Xarray DataArray")
+    try:
+        return XTensorConstant(
+            XTensorType(dtype=x.dtype, dims=x.dims, shape=x.shape),
+            x.values.copy(),
+            name=name,
+        )
+    except TypeError:
+        raise TypeError(f"Could not convert {x} to XTensorType")
+
+
+if XARRAY_AVAILABLE:
+
+    @_as_symbolic.register(xr.DataArray)
+    def as_symbolic_xarray(x, **kwargs):
+        return xtensor_constant(x, **kwargs)
+
+
+def as_xtensor_variable(x, name=None):
```

Comment:
would this have dims?

Reply:
This is analogous to `as_tensor_variable` which doesn't allow you to pass constructor information. Right now only accepts xarrays

Reply:
There is an Op `xtensor_from_tensor` that can be used to add dims to a vanilla tensor that only has shape